executor/new_join.go

```diff
@@ -0,0 +1,253 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"bytes"
+	"sort"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/util/codec"
+	"github.com/pingcap/tidb/util/types"
+)
+
+type orderedRow struct {
+	key []byte
+	row *Row
+}
+
+type orderedRows []orderedRow
+
+// Len returns the number of rows.
+func (e orderedRows) Len() int {
+	return len(e)
+}
+
+// Swap implements sort.Interface Swap interface.
+func (e orderedRows) Swap(i, j int) {
+	e[i], e[j] = e[j], e[i]
+}
+
+// Less implements sort.Interface Less interface.
+func (e orderedRows) Less(i, j int) bool {
+	return bytes.Compare(e[i].key, e[j].key) < 0
+}
+
+// IndexLookUpJoin fetches batches of data from outer executor and constructs ranges for inner executor.
+type IndexLookUpJoin struct {
```

Comment:
I think the exec suffix is unnecessary.  The type is executor and there are no name conflicts with it